In [2]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

In [4]:
# Train Phase transformations
train_transforms = transforms.Compose([
                                       transforms.Resize((28, 28)),
                                       transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,)) # The mean and std have to be sequences (e.g., tuples), therefore you should add a comma after the values.
                                       # Note the difference between (0.1307) and (0.1307,)
                                       ])

# Test Phase transformations
test_transforms = transforms.Compose([
                                      #  transforms.Resize((28, 28)),
                                      #  transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])


In [5]:
train = datasets.MNIST('./data', train=True, download=True, transform=train_transforms)
test = datasets.MNIST('./data', train=False, download=True, transform=test_transforms)

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 9.91M/9.91M [00:02<00:00, 4.89MB/s]


Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 28.9k/28.9k [00:00<00:00, 157kB/s]


Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 1.65M/1.65M [00:01<00:00, 1.27MB/s]


Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████| 4.54k/4.54k [00:00<00:00, 3.79MB/s]

Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw



In [6]:
SEED = 1

# CUDA?
cuda = torch.cuda.is_available()
print("CUDA Available?", cuda)

# For reproducibility
torch.manual_seed(SEED)

if cuda:
    torch.cuda.manual_seed(SEED)

# dataloader arguments - something you'll fetch these from cmdprmt
dataloader_args = dict(shuffle=True, batch_size=128, num_workers=4, pin_memory=True) if cuda else dict(shuffle=True, batch_size=64)

# train dataloader
train_loader = torch.utils.data.DataLoader(train, **dataloader_args)

# test dataloader
test_loader = torch.utils.data.DataLoader(test, **dataloader_args)

CUDA Available? True


In [7]:
class Model4(nn.Module):
    def __init__(self):
        super(Model4, self).__init__()

        self.convblock1 = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT = 28*28 | OUTPUT 26*26 | RF 3

        self.convblock2 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 26*26 | OUTPUT 24*24 | RF 5

        self.convblock3 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 24*24 | OUTPUT 22*22 | RF 7

        self.pool1 = nn.MaxPool2d(2, 2) # INPUT 22*22 | OUTPUT 11*11 | RF 8

        self.convblock4 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 11*11 | OUTPUT 9*9 | RF 12

        self.convblock5 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=20, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(20),
            nn.ReLU()
        ) # INPUT 9*9 | OUTPUT 7*7 | RF 16


        self.convblock6 = nn.Sequential(
            nn.Conv2d(in_channels=20, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 7*7 | OUTPUT 5*5 | RF 20

        self.convblock7 = nn.Sequential(
            nn.Conv2d(in_channels=10, out_channels=10, kernel_size=(3, 3), padding=0),
            nn.BatchNorm2d(10),
            nn.ReLU()
        ) # INPUT 5*5 | OUTPUT 3*3 | RF 24

        self.gap = nn.Sequential(
            nn.AvgPool2d(3)
        ) # INPUT 3*3 | OUTPUT 1*1 | RF 28

        self.dropout = nn.Dropout(0.1)

    def forward(self, x):
        x = self.convblock1(x)
        x = self.convblock2(x)
        #x = self.dropout(x)
        x = self.convblock3(x)
        x = self.pool1(x)
        x = self.convblock4(x)
        x = self.convblock5(x)
        x = self.dropout(x)
        x = self.convblock6(x)
        x = self.convblock7(x)
        x = self.gap(x)
        x = x.view(-1, 10)
        return F.log_softmax(x, dim=-1)

In [8]:
%pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print(device)
model = Model4().to(device)
summary(model, input_size=(1, 28, 28))

Note: you may need to restart the kernel to use updated packages.
cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 10, 26, 26]             100
       BatchNorm2d-2           [-1, 10, 26, 26]              20
              ReLU-3           [-1, 10, 26, 26]               0
            Conv2d-4           [-1, 10, 24, 24]             910
       BatchNorm2d-5           [-1, 10, 24, 24]              20
              ReLU-6           [-1, 10, 24, 24]               0
            Conv2d-7           [-1, 10, 22, 22]             910
       BatchNorm2d-8           [-1, 10, 22, 22]              20
              ReLU-9           [-1, 10, 22, 22]               0
        MaxPool2d-10           [-1, 10, 11, 11]               0
           Conv2d-11             [-1, 10, 9, 9]             910
      BatchNorm2d-12             [-1, 10, 9, 9]              20
             ReLU-13            

In [9]:
from tqdm import tqdm

train_losses = []
test_losses = []
train_acc = []
test_acc = []

def train(model, device, train_loader, optimizer, epoch):
  model.train()
  pbar = tqdm(train_loader)
  correct = 0
  processed = 0
  for batch_idx, (data, target) in enumerate(pbar):
    # get samples
    data, target = data.to(device), target.to(device)

    # Init
    optimizer.zero_grad()
    # In PyTorch, we need to set the gradients to zero before starting to do backpropragation because PyTorch accumulates the gradients on subsequent backward passes.
    # Because of this, when you start your training loop, ideally you should zero out the gradients so that you do the parameter update correctly.

    # Predict
    y_pred = model(data)

    # Calculate loss
    loss = F.nll_loss(y_pred, target)
    train_losses.append(loss)

    # Backpropagation
    loss.backward()
    optimizer.step()

    # Update pbar-tqdm

    pred = y_pred.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
    correct += pred.eq(target.view_as(pred)).sum().item()
    processed += len(data)

    pbar.set_description(desc= f'Loss={loss.item()} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')
    train_acc.append(100*correct/processed)

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    test_losses.append(test_loss)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

    test_acc.append(100. * correct / len(test_loader.dataset))

In [10]:
from torch.optim.lr_scheduler import StepLR

model =  Model4().to(device)
optimizer = optim.Adam(model.parameters(), lr=0.001) #learning rate reduced to 0.001 from 0.01
scheduler = StepLR(optimizer, step_size=8, gamma=0.1)
EPOCHS = 15

for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)
    scheduler.step()


EPOCH: 0


Loss=0.2490364909172058 Batch_id=468 Accuracy=93.12: 100%|██████████| 469/469 [00:10<00:00, 44.72it/s] 



Test set: Average loss: 0.2375, Accuracy: 9816/10000 (98.16%)

EPOCH: 1


Loss=0.15466056764125824 Batch_id=468 Accuracy=98.00: 100%|██████████| 469/469 [00:07<00:00, 66.28it/s]



Test set: Average loss: 0.1369, Accuracy: 9848/10000 (98.48%)

EPOCH: 2


Loss=0.16827517747879028 Batch_id=468 Accuracy=98.43: 100%|██████████| 469/469 [00:07<00:00, 66.85it/s] 



Test set: Average loss: 0.0837, Accuracy: 9880/10000 (98.80%)

EPOCH: 3


Loss=0.09689375758171082 Batch_id=468 Accuracy=98.64: 100%|██████████| 469/469 [00:07<00:00, 66.48it/s] 



Test set: Average loss: 0.0666, Accuracy: 9890/10000 (98.90%)

EPOCH: 4


Loss=0.10837340354919434 Batch_id=468 Accuracy=98.68: 100%|██████████| 469/469 [00:07<00:00, 65.74it/s] 



Test set: Average loss: 0.0511, Accuracy: 9912/10000 (99.12%)

EPOCH: 5


Loss=0.06592915207147598 Batch_id=468 Accuracy=98.88: 100%|██████████| 469/469 [00:07<00:00, 66.26it/s] 



Test set: Average loss: 0.0471, Accuracy: 9908/10000 (99.08%)

EPOCH: 6


Loss=0.10578694194555283 Batch_id=468 Accuracy=98.84: 100%|██████████| 469/469 [00:07<00:00, 66.98it/s] 



Test set: Average loss: 0.0488, Accuracy: 9902/10000 (99.02%)

EPOCH: 7


Loss=0.03217274695634842 Batch_id=468 Accuracy=99.00: 100%|██████████| 469/469 [00:06<00:00, 67.42it/s] 



Test set: Average loss: 0.0386, Accuracy: 9921/10000 (99.21%)

EPOCH: 8


Loss=0.024886390194296837 Batch_id=468 Accuracy=99.27: 100%|██████████| 469/469 [00:06<00:00, 68.01it/s]



Test set: Average loss: 0.0269, Accuracy: 9943/10000 (99.43%)

EPOCH: 9


Loss=0.050524670630693436 Batch_id=468 Accuracy=99.28: 100%|██████████| 469/469 [00:06<00:00, 67.43it/s]



Test set: Average loss: 0.0273, Accuracy: 9938/10000 (99.38%)

EPOCH: 10


Loss=0.027504606172442436 Batch_id=468 Accuracy=99.34: 100%|██████████| 469/469 [00:07<00:00, 66.31it/s]



Test set: Average loss: 0.0258, Accuracy: 9940/10000 (99.40%)

EPOCH: 11


Loss=0.034093741327524185 Batch_id=468 Accuracy=99.31: 100%|██████████| 469/469 [00:07<00:00, 66.67it/s]



Test set: Average loss: 0.0257, Accuracy: 9946/10000 (99.46%)

EPOCH: 12


Loss=0.024683421477675438 Batch_id=468 Accuracy=99.38: 100%|██████████| 469/469 [00:07<00:00, 66.29it/s]



Test set: Average loss: 0.0249, Accuracy: 9948/10000 (99.48%)

EPOCH: 13


Loss=0.08600360155105591 Batch_id=468 Accuracy=99.36: 100%|██████████| 469/469 [00:06<00:00, 67.06it/s] 



Test set: Average loss: 0.0247, Accuracy: 9948/10000 (99.48%)

EPOCH: 14


Loss=0.05660015344619751 Batch_id=468 Accuracy=99.39: 100%|██████████| 469/469 [00:06<00:00, 67.24it/s] 



Test set: Average loss: 0.0244, Accuracy: 9947/10000 (99.47%)



TARGET:

Keep the total parameter count less than 8000

Reach test accuracy > 99.4 in less than 15 epochs

RESULT:

Parameters : 7,530

Best Training Accuracy : 99.39

Best Test Accuracy : 99.48

ANALYSIS:

Model is underfitting

Test accuracy has reached 99.4 and is consistent from 10th epoch

Adding StepLR helped the model to perform better